www/source/blog/2017-10-15-Configuration-Plans.html.md

```diff
@@ -0,0 +1,119 @@
+anticipated. So what to do? We could try and add templates to
+support all possible uses, but that will rapidly become untenable. What
+good packages should do instead is provide well documented entry points, so
+that consumers of our package can easily supply their own configuration
+in a new plan -- a Configuration Plan.
+
+## Building a Configuration Plan
+
+Let's have a look at an example. There are actually a fair number of
+live examples in the habitat repository - check out the [Builder API
+Proxy](https://github.com/habitat-sh/habitat/tree/master/components/builder-api-proxy), which is a Configuration Plan for nginx. But for now, let's build our own.
+
+I've created a tiny binary that simply reads a file and prints it out.
+You can [see the
+source](https://github.com/thommay/configuration_plan_demo), and you can
+try running it by entering a studio and running the demo.
+
+```bash
+$ hab studio enter
+(studio) $ hab svc start thom/configuration_plan_demo
+(studio) $ sup-log
+```
+
+You'll see the demo print out the configuration file name, and then the
+contents of the file.
+
+```
+configuration_plan_demo.default(O): Filename is: /hab/svc/configuration_plan/config/output.txt
+configuration_plan_demo.default(O): The original plan
+
+### Making a new plan
+Now, let's create a Configuration Plan that provides a new file for the
+demo. First, we'll create a new plan:
+
+```bash
+$ hab plan init -o <your_origin> configuration_plan
+$ cd configuration_plan
+```
+
+We now have a `configuration_plan` directory, containing a `plan.sh`,
+a `config` dir, and a `hooks` dir.
+
+Let's create that configuration file!
+
+```bash
+$ echo 'a really great demo' > config/my_configuration_plan.txt
+```
+
+Open up your `plan.sh`, and replace it with the following, being sure to
+set your origin correctly.
+
+```bash
+pkg_origin=your_origin
+pkg_name=configuration_plan
+pkg_version=0.1.0
+pkg_deps=(thom/configuration_plan_demo)
+pkg_svc_run="configuration_plan_demo ${pkg_svc_config_path}/my_configuration_plan.txt"
+
+do_build() {
+  return 0
+}
+
+do_install() {
+  return 0
+}
+
+do_unpack() {
+  return 0
+}
+```
+
+Because we're shipping nothing but configuration, we tell Habitat that
+we don't need to unpack, build or install. Instead, we simply depend on
+`thom/configuration_plan_demo`, and then ensure that we use our new
+configuration file.
+
+### Running the plan
+Now we can build and run our plan. Enter a studio as usual, run `build`
+to build your plan, and then install it. On my machine, I ran:
+
+```bash
+$ hab studio enter
+(studio) $ build
+(studio) $ hab pkg install /src/results/thom-configuration_plan-0.1.0-20171016162630-x86_64-linux.hart
+(studio) $ hab svc start thom/configuration_plan
+```
+
+By running `sup-log`, we should see our demo start and use the new
+config file.
+
+```
+configuration_plan.default(O): Filename is: /hab/svc/configuration_plan/config/my_configuration_plan.txt
+configuration_plan.default(O): a really great demo
```

Comment:
I think that this simplified demo begs the question, why wouldn't I ask the template maintainer to add a template variable for this line of text. It definitely works as a demo as you are pointing to a different file path in the service run operation in the plan and that is not something that you can override. But in some ways the pointing to a different file almost, without the configuration defaults, doesn't feel like your are updating configuration.

I guess what I'm trying to say is that it feels almost a little too trivial.

Reply:
Could it remain trivial and at the end here some more examples could be provided like you mentioned above (like the nginx)?

I know that I have used this pattern when the `core/haproxy` package was not doing what I wanted and then configuration was broken for me at the time of using it (https://github.com/learn-chef/hab-two-tier/blob/master/haproxy/config/haproxy.conf). I also did this with`core/httpd` (https://github.com/learn-chef/hab-two-tier/tree/master/webapp) because the default behavior was not working correctly.

> The httpd example is a pretty gnarly one. If you look within the conf I use the `pkgPathFor "core/httpd"` for the ServerRoot because the original httpd requires that you have a few directories. It was either that or make a bunch of symlinks.